Add SpriteBatch API

erf opened the pull request:
This is an PR for #282.

- added a `SpriteBatch` class which buffers rects, transforms and optional colors, and renders using `Canvas.drawAtlas`
- added a `SpriteBatchComponent` which accepts a `SpriteBatch`
- added an Example which render a bunch of sprites in various ways ( using `SpriteBatch` )

EDIT: 
- added new doc entry `SpriteBatch` under `images.md`
- updated CHANGELOG

erf commented:
Ok, i will look into those changes. 

Here is a screenshot:
![Simulator Screen Shot - iPhone 11 Pro - 2020-04-16 at 14 32 01](https://user-images.githubusercontent.com/1562523/79512420-b1a79180-8041-11ea-9850-44b432b9ce00.png)

erf commented:
I added a doc entry and updated changelog.

erickzanardo commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
 ## [next]
 
+## 0.20.0
```
Leave this on the `[next]` entry, which version will be is defined when we release the version.

erickzanardo commented on doc/images.md:
```diff
@@ -45,6 +45,18 @@ Sprites can also be used as widgets, to do so, just use `Flame.util.spriteAsWidg
+
+An `SpriteBatchComponent` is also available for your convenience.
+
+See example [here](examples/sprite_batch).
```
Does this link work? I think it should be `docs/example/sprite_batch` doesn't it?

erf replied:
I think you are right. I looked at `examples/animation_widget` which is also wrong then? Will update that also.

erickzanardo commented on lib/sprite_batch.dart:
```diff
@@ -0,0 +1,82 @@
+
+  SpriteBatch.fromImage(this.atlas);
+
+  int get width => atlas.width;
```
Just to make sure, I would verify for null here

Example
```dart
  int get width => atlas?.width;
```

Same thing for `height` and `size`

erf replied:
do `atlas?.height ?? 0` or `atlas?.height` and return `null`

for size do:
```
  Size get size => atlas == null ? null : Size(width.toDouble(), height.toDouble());
```
or return size with 0 values ?

btw: i think i would prefer to pass an image to the constructor, which is not null, so we dont have to do async load stuff and null checks there. assets could be loaded outside and prepared beforehand

erickzanardo replied:
I think we should just return null, not 0 as default, just to avoid a NPE inside the method.

erf replied:
In the latest commit you dont have to check for null

erf commented:
Should i amend and force push changes as i fix them here?

erf commented:
Rebased and force pushed changes with fixes.

luanpotter commented on doc/examples/sprite_batch/lib/main.dart:
```diff
@@ -0,0 +1,65 @@
+
+    add(SpriteBatchComponent.fromSpriteBatch(
+      spriteBatch,
+      //paint: Paint()..invertColors = true,
```
please either uncomment or remove

luanpotter commented on doc/images.md:
```diff
@@ -43,7 +43,19 @@ All render methods from the Sprite class can receive a `Paint` instance on the o
+
+## SpriteBatch
+
+If you have an image atlas ( an image with smaller images inside ), and would like to render it effectively - `SpriteBatch` does the job.
```
1. please don't add spaces inside parenthesis (like this).
2. we pretty much call "sprite atlas" spritesheets everywhere else. AFAICS this is the same thing just being rendered more efficiently. so maybe I'd change the text to be
> If you have a sprite sheet (also called an image atlas, which is ...), ...
just to make it more clear that spritesheet and atlas from the flame perspective are the same thing, this is just a new way to render it (correct me if I'm wrong)

erf replied:
Yeah, they are essentially doing the same thing, only SpriteSheet now has a bit more functionality. Should replace "SpriteSheet" with "SpriteBatch" later ( and perhaps rename SpriteBatch to SpriteSheet again )

luanpotter commented on doc/images.md:
```diff
@@ -43,7 +43,19 @@ All render methods from the Sprite class can receive a `Paint` instance on the o
+
+If you have an image atlas ( an image with smaller images inside ), and would like to render it effectively - `SpriteBatch` does the job.
+
+Give it the filename of the image, and then add rectangles which describes various part of the image, in addition to transforms ( position, scale and rotation ) and optional colors.
```
(spaces in parenthesis)

luanpotter commented on lib/components/sprite_batch_component.dart:
```diff
@@ -0,0 +1,36 @@
+
+  @override
+  bool loaded() {
+    return spriteBatch != null && spriteBatch.atlas != null;
```
you can do `spriteBatch?.atlas != null`. and then you can inline this function like this
```
bool loaded() => spriteBatch?.atlas != null;
```

erf replied:
could `loaded` be a `get` property here?

erf commented:
pushed changes

erf commented:
I added a new commit where i changed the constructor, given the image asset, and created an async factory method which creates and return a new `SpriteBatch` with a preloaded asses, as i commented in https://github.com/flame-engine/flame/issues/289#issuecomment-615374515 
In this way we always guarantee that the atlas image is loaded and we dont have to do null checks.

erf commented:
Btw i wanted to pass optional named parameters for rects, transforms and colors to the spritebatch, but they become null, even if i declared them as empty lists in the class. Not sure why this is. I added a question on stackoverflow here:
https://stackoverflow.com/questions/61303278/optional-constructor-parameter-set-the-list-to-null-in-dart/61303489?noredirect=1#comment108449447_61303489

but i think it might be better to use named constructors at a later point, in order to create sprite sheets etc.

luanpotter commented on lib/components/sprite_batch_component.dart:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  @override
+  bool loaded() => true;
```
this returns true by default, no need to override it

erf commented:
Ok, should be good to go.

renancaraujo commented:
This is amazing. Good to go!
